### PR TITLE
feat: zcap - adds caveat expiry support.

### PR DESCRIPTION
**Pull request by soluchok**
Signed-off-by: Andrii Soluk <isoluchok@gmail.com>

### CODE REVIEW

File: pkg/zcapld/create.go [hunk 1]
```diff
@@ -44,6 +44,18 @@ type CapabilityOptions struct {
...
+	Caveats          []Caveat
+}
+
+// Caveat types.
+const (
+	CaveatTypeExpiry = "expiry"
+)
+
+// Caveat is a restriction on how capability may be used.
+type Caveat struct {
```
Comment (llorllale) [inline pkg/zcapld/create.go:56]: I think caveats should be a generic mechanism - there will be several types of caveats, not just expiry.

Can you please leave a TODO to generalize this mechanism later?

Comment (soluchok) [reply pkg/zcapld/create.go:56]: ok, I will. But it seems like we just need to extend a structure.
e.g
```
type Caveat struct {
	Type     string `json:"type"`
	Duration uint64 `json:"duration"`
        URI uint64 `json:"uri"`
}
```

Comment (soluchok) [reply pkg/zcapld/create.go:57]: @llorllale Done!

Comment (llorllale) [reply pkg/zcapld/create.go:57]: > ok, I will. But it seems like we just need to extend a structure.

Which is why I think a better way would be to base this on maps. The only common property will be "type". (and "id" if we were to model this strictly in JSONLD terms, but I have other thoughts about that...)

File: pkg/zcapld/verify.go [hunk 2]
```diff
@@ -193,7 +194,14 @@ func (v *Verifier) verifyCapabilityChain(
 	}
 
 	// 4.2. Ensure that the caveats are met on the root capability.
-	// TODO verify caveats
+	for i := range capability.Caveats {
+		if capability.Caveats[i].Type == CaveatTypeExpiry {
```
Comment (llorllale) [inline pkg/zcapld/verify.go:198]: This is the reason why I think the mechanism should be generalized: users should be able to inject their own caveat validators.

**Review by llorllale — APPROVED** [review]

File: pkg/zcapld/verify.go (hunk 2 above)
Comment (soluchok) [reply pkg/zcapld/verify.go:198]: got it

**Review by aholovko — APPROVED** [review]

**Pull request merged**